Добаления компонентов модального окна для ингредиентов

Johnny-Silverhand2077 opened the pull request:
добавление разметки в файле App

Johnny-Silverhand2077 committed:
Добаления компонентов модального окна для ингредиентов
добавление разметки в файле App

Johnny-Silverhand2077 committed:
Добавление конструкции слвйса бургера
добавление слайсов userSlice, FeedDataSlice, userOrderHistorySlice
добавление фукциональности заказа
добавление функции выхода из системы и добавления заказов

Johnny-Silverhand2077 committed:
Добаление компонента selector
добавление функций в файл FeedDataSlice
исправлен баг в изображении ленты заказов

gennady-bars commented on .env.example:
```diff
@@ -1 +0,0 @@
-BURGER_API_URL=https://norma.nomoreparties.space/api
```
* В проекте по инструкции должен быть файл `.env.example`, где указан базовый урл для запросов, чтобы ревьюеру не искать его https://disk.yandex.ru/i/6P4U2CxEJKsOsw

gennady-bars commented on src/components/app-header/app-header.tsx:
```diff
@@ -1,4 +1,15 @@
 import { AppHeaderUI } from '@ui';
+import { selectUser } from '../../slices/UserInfoSlice';
+import { useSelector } from '../../services/store';
 
```
**Можно лучше**


Если интересно, то можно создать **универсальную** функцию запроса `request` , которая будет содержать в себе уже **базовый урл**, проверку на **ok** и проверку на `success`, чтобы не дублировать их в каждом запросе.

```js

// 1 раз объявляем базовый урл
export const BASE_URL = "https://norma.nomoreparties.space/api/";

// создаем функцию проверки ответа на `ok`
const checkResponse = (res) => {
  if (res.ok) {
    return res.json();
  }
  // не забываем выкидывать ошибку, чтобы она попала в `catch`
  return Promise.reject(`Ошибка ${res.status}`);
};

// создаем функцию проверки на `success`
const checkSuccess = (res) => {
  if (res && res.success) {
    return res;
  }
  // не забываем выкидывать ошибку, чтобы она попала в `catch`
  return Promise.reject(`Ответ не success: ${res}`);
};

// создаем универсальную фукнцию запроса с проверкой ответа и `success`
// В вызов приходит `endpoint`(часть урла, которая идет после базового) и опции
const request = (endpoint, options) => {
  // а также в ней базовый урл сразу прописывается, чтобы не дублировать в каждом запросе
  return fetch(`${BASE_URL}${endpoint}`, options)
    .then(checkResponse)
    .then(checkSuccess);
};

```


И теперь не нужно будет дублировать все это в каждом запросе. Нужно просто заменить в запросах все `fetch` на `request`, удалив лишние проверки и базовый урл. Все остальное останется без изменений

Код станет намного чище. Будет мелкий аналог `axios`

Пример запроса ингредиентов

```js
const getIngredients =  () => request("ingredients");

```

В `get`-запросах даже не нужно указывать 2й параметр. Он получится `undefined`, что без проблем для `fetch` (который внутри).

gennady-bars commented on src/slices/IngredientsSlice.ts:
```diff
@@ -0,0 +1,52 @@
+  error: null
+};
+
+export const getIngredients = createAsyncThunk(
+  'ingredients/getIngredients',
+  async () => {
+    const res = await getIngredientsApi();
+    return res;
+  }
```
**Можно лучше**

* Если внутри **безымянной** (стрелочной) функции вызывается одна функция с **точно такими же** аргументами, то эта безымянная функция не нужна. Это лишняя обертка вокруг самой функции, которую можно использовать просто сразу в коде


```js
export const getIngredients = createAsyncThunk(
  'ingredients/getIngredients',
  getIngredientsApi
);
```

gennady-bars commented on src/pages/login/login.tsx:
```diff
@@ -1,13 +1,30 @@
+import {useSelector} from '../../services/store' 
 
 export const Login: FC = () => {
   const [email, setEmail] = useState('');
   const [password, setPassword] = useState('');
+  const dispatch = useDispatch()
```
**Можно лучше**

Можно сделать универсальный **кастомный** хук для контроля любого количества инпутов в любых формах:

```js
export function useForm(inputValues={}) {
  const [values, setValues] = useState(inputValues);

  const handleChange = (event) => {
    const {value, name} = event.target;
    setValues({...values, [name]: value});
  };
  return {values, handleChange, setValues};
}

```

Этот код помещают в отдельный файл `useForm.js` в папке `hooks` и импортируют функцию туда, где нужно контролировать инпуты

И Вам не нужно будет теперь вручную создавать функции обработки инпутов и т д. Все будет в одной строчке кода:

```jsx
  const {values, handleChange, setValues} = useForm({});

```

gennady-bars commented on src/components/app/app.tsx:
```diff
@@ -1,14 +1,144 @@
+  const navigate = useNavigate();
+  const locationState = location.state as { background?: Location };
+  const background =
+    locationState && location.state?.background;
```
* Отлично сделана логика отображения попапа и страницы  под ним

gennady-bars reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)

Посмотрите на **гитхабе** все комментарии к коду (нужно прокрутить вниз страницу там)

Работа проделана огромная:




* Отлично, что в `prepare` создаете `id`
* Отлично сделана логика отображения попапа и страницы  под ним
* Хорошо сделали `ProtectedRoute`


но есть некоторые недочеты:



* В проекте по инструкции должен быть файл `.env.example`, где указан базовый урл для запросов, чтобы ревьюеру не искать его https://disk.yandex.ru/i/6P4U2CxEJKsOsw 
* Не открывается попап в истории заказов https://disk.yandex.ru/i/azJp_Lv3tq5Tpw 


**Можно лучше**

* при клике на **логотип** обычно идет переход на **главную** страницу
* Если внутри **безымянной** (стрелочной) функции вызывается одна функция с **точно такими же** аргументами, то эта безымянная функция не нужна. Это лишняя обертка вокруг самой функции, которую можно использовать просто сразу в коде

Johnny-Silverhand2077 committed:
Испрааление замечаний ревью
исправлени баг при открытии модального окна в профиле, история заказов
исправленно замечание ревью при отрытом файле .env.example

gennady-bars reviewed (approved):
Поздравляю! Ваша работа принята.

Вы отлично потрудились.

Удачного дальнейшего обучения.

Johnny-Silverhand2077 committed:
Добавление тестов cypress и jest
созданны 4 файла теста для работы слайсов
